src/clients/compositor/sway.rs

```
@@ -21,36 +37,59 @@ impl Client {
         let client = Arc::new(Mutex::new(Connection::new().await?));
         info!("Sway IPC subscription client connected");
 
-        let (workspace_tx, workspace_rx) = channel(16);
+        Ok(Self {
+            client,
+            task_state: Mutex::new(TaskState {
+                listeners: Arc::new(Vec::new()),
+                join_handle: None,
+            }),
+        })
+    }
+
+    pub async fn add_listener(&self, event_type: EventType, f: Box<SyncFn>) -> Result<()> {
+        // abort current running task
+        let TaskState {
+            join_handle,
+            listeners,
+        } = &mut *self.task_state.lock().await;
+
+        if let Some(handle) = join_handle.take() {
+            handle.abort();
+            let _ = handle.await;
+        }
 
-        {
-            // create 2nd client as subscription takes ownership
-            let client = Connection::new().await?;
-            let workspace_tx = workspace_tx.clone();
+        // Only the task and self have a reference to listeners, and we just abort the task. This
+        // is the only reference to listeners, so we can safely unwrap.
+        let listeners_mut = Arc::get_mut(listeners).unwrap();
```

`.unwrap` should be [avoided](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap) in prod code. In part, it's because it is unclear that the situation *is* safe, without relying on a comment above. Generally if you're unwrapping because the code would have to get into an "impossible" state you should use `expect` instead.

In this case since you're returning a `Result` anyway, it would be better to return an error either with `?` or explicitly, and handle the situation at the call.